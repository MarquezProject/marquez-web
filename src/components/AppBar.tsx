import React, { ReactElement } from 'react'
import { Link } from 'react-router-dom'
import { AppBar, Toolbar, Typography } from '@material-ui/core'
import {
  withStyles,
  createStyles,
  WithStyles as IWithStyles,
  Theme as ITheme
} from '@material-ui/core/styles'
import Box from '@material-ui/core/Box'
import 'typeface-rajdhani'
import '../img/we-logo.png'

const styles = ({ spacing, zIndex }: ITheme) => {
  return createStyles({
    appBar: {
      zIndex: zIndex.drawer + 1
    },
    rightToolbar: {
      marginLeft: 'auto',
      marginRight: 2.5
    },
    feedbackButton: {
      padding: spacing(1),
      backgroundColor: '#2B2B33',
      color: 'white',
      fontSize: 15,
      borderWidth: 1.2,
      '&:hover': {
        color: '#71ddbf',
        cursor: 'pointer',
        borderColor: '#71ddbf'
      }
    }
  })
}

interface IProps extends IWithStyles<typeof styles> {}

const MyAppBar = (props: IProps): ReactElement => {
  const { classes } = props
  return (
    <AppBar position='fixed' className={classes.appBar}>
      <Toolbar>
        <Box mr={2}>
          <img src='img/we-logo.png' height={50} alt='WeWork Logo' />
        </Box>
        <Typography
          variant='h4'
          color='inherit'
          noWrap
          style={{ fontFamily: 'rajdhani', fontWeight: 'bold' }}
        >
          <Link to='/' className='link'>
            MARQUEZ
          </Link>
        </Typography>
        <div className={classes.rightToolbar}>
          <a href='https://forms.gle/gWj8fn8iYiTsKEPD6' target='_blank' rel='noopener noreferrer'>
            <button className={classes.feedbackButton}> Feedback </button>
          </a>
        </div>
      </Toolbar>
    </AppBar>
  )
}

export default withStyles(styles)(MyAppBar)
