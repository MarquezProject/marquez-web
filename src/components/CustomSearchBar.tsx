import React, { FunctionComponent, useState } from 'react'
import SearchBar from 'material-ui-search-bar'
import { findMatchingEntities } from '../actionCreators'
import {
  withStyles,
  createStyles,
  WithStyles as IWithStyles,
  Theme as ITheme
} from '@material-ui/core/styles'

import { useHistory } from "react-router-dom";

interface IProps {
  findMatchingEntities: typeof findMatchingEntities
  setShowJobs: (bool: boolean) => void
  showJobs: boolean
}

type IAllProps = IWithStyles<typeof styles> & IProps

const styles = (_theme: ITheme) => {
  return createStyles({
    search: {
      position: 'fixed',
      bottom: '52vh',
      width: '90%',
      left: '5%',
      zIndex: 4
    }
  })
}

const CustomSearchBar: FunctionComponent<IAllProps> = props => {
  const [search, setSearch] = useState("")
  const { classes } = props
  let history = useHistory();

  const searchChanged = (searchString: string) => {
    setSearch(searchString)
    props.findMatchingEntities(searchString)
    searchString == '' ? props.setShowJobs(false) : props.setShowJobs(true)
  }

  const cancelledSearch = () => {
    setSearch('')
    searchChanged('')
  }

  const onRequestSearch = () => {
    history.push("/");
  }

  return (
    <SearchBar
      className={classes.search}
      value={search}
      onChange={searchChanged}
      onCancelSearch={cancelledSearch}
      onRequestSearch={onRequestSearch}
    />
  )
}


export default withStyles(styles)(CustomSearchBar)
