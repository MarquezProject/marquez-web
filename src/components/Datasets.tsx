import React, { ReactElement } from 'react'
import * as RRD from 'react-router-dom'
import { Typography, Box } from '@material-ui/core'

interface IProps {}

interface IState {}

type IAllProps = RRD.RouteComponentProps & IProps

class Datasets extends React.Component<IAllProps, IState> {
  render(): ReactElement {
    return (
      <Box mt={10}>
        <Typography color='primary' align='center'>
          {' '}
          Datasets Component{' '}
        </Typography>
      </Box>
    )
  }
}

export default Datasets
