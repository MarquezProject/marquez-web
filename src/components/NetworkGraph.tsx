import * as React from 'react'
import * as d3 from 'd3'

import {
  withStyles,
  createStyles,
  WithStyles as IWithStyles,
  Theme
} from '@material-ui/core/styles'

import Legend from './Legend'

import { IDataset, IJob } from '../types/'

import _find from 'lodash/find'
import _filter from 'lodash/filter'
import _flatten from 'lodash/flatten'
import _map from 'lodash/map'
import _sortBy from 'lodash/sortBy'

import { select, event } from 'd3-selection'
import { hierarchy, tree } from 'd3-hierarchy'
import { linkHorizontal } from 'd3-shape'
import { drag } from 'd3-drag'
import { zoom } from 'd3-zoom'

import Loader from './Loader'
import { IJobRunAPI } from '../types/api'

const globalStyles = require('../global_styles.css')
const { jobRunNew, jobRunFailed, jobRunCompleted, jobRunAborted, jobRunRunning } = globalStyles

const colorMap = {
  NEW: jobRunNew,
  FAILED: jobRunFailed,
  COMPLETED: jobRunCompleted,
  ABORTED: jobRunAborted,
  RUNNING: jobRunRunning
}

const styles = ({ palette }: Theme) => {
  return createStyles({
    networkBackground: {
      background: palette.common.black,
      width: '100%',
      height: '50vh',
      position: 'fixed',
      display: 'flex',
      alignItems: 'center',
      zIndex: 2
    },
    tooltip: {
      position: 'absolute',
      visibility: 'hidden',
      background: '#fff',
      color: palette.primary.main,
      zIndex: 3,
      padding: '5px 10px',
      font: '12px sans-serif',
      borderRadius: '2px',
      pointerEvents: 'none',
      opacity: 0.8
    },
    legend: {
      position: 'fixed',
      bottom: '59vh',
      right: '6%',
      zIndex: 3
    }
  })
}

interface IProps {
  jobs: IJob[]
  datasets: IDataset[]
  isLoading: boolean
  router: any
}

type IAllProps = IWithStyles<typeof styles> & IProps

export class NetworkGraph extends React.Component<IAllProps, {}> {
  shouldComponentUpdate(newProps: IProps) {
    const urlBreakdown = newProps.router.location.pathname.split('/')
    const nodeId = urlBreakdown[2]

    const svg: d3.Selection<SVGElement, void, HTMLElement, void> = select('#network-graph')

    if (svg.empty()) {
      return true
    }
    const height = +svg.style('height').replace('px', '')
    const width = +svg.style('width').replace('px', '')

    const isDataset = (node: any) => {
      const name = node.name || node.data.name
      return _find(newProps.datasets, d => d.name == name)
    }

    const findChildren = (node: any) => {
      let children
      if (isDataset(node)) {
        children = _filter(newProps.jobs, j => j.inputs.includes(node.name))
      } else {
        const job = _find(newProps.jobs, j => j.name == node.name)
        children = job ? _filter(newProps.datasets, d => job.outputs.includes(d.name) && !job.inputs.includes(d.name)) : []
      }
      return children
    }

    const findParents = (node: any) => {
      let parents
      if (isDataset(node)) {
        parents = _filter(newProps.jobs, j => j.outputs.includes(node.name))
      } else {
        const job = _find(newProps.jobs, j => j.name == node.name)
        parents = job ? _filter(newProps.datasets, j => job.inputs.includes(j.name) && !job.outputs.includes(j.name)) : []
      }
      return parents
    }

    const getLineages = () => {
      const searchedDatasets = _filter(newProps.datasets, d => d.matches)
      const searchedJobs = _filter(newProps.jobs, j => j.matches)
      const allNodes = [...searchedDatasets, ...searchedJobs]

      const lineages = _map(allNodes, (rootNode: any) => {
        rootNode.children = findChildren(rootNode)
        let children = rootNode.children
        while (children.length > 0) {
          _map(children, child => {
            child.children = findChildren(child)
          })
          children = _flatten(_map(children, chi => chi.children))
        }
        return rootNode
      })
      return lineages
    }

    const getReverseLineage = (node: any) => {
      node.children = findParents(node)
      let children = node.children
      while (children.length > 0) {
        _map(children, child => {
          child.children = findParents(child)
        })
        children = _flatten(_map(children, chi => chi.children))
      }
      return node
    }

    const circleHighlight = '#ffffff'
    const linkHighlight = '#b0b0b0'
    const defaultHighlight = '#575757'
    const labelHighlight = '#ffffff'
    const radius = 8
    const square = 13
    const strokeWidth = 5

    function findJobColor(job: any) {
      console.log('Job: ', job)
      const key = job.data.latestRun.runState as IJobRunAPI['runState']
      const color = colorMap[key]
      return color
    }

    function graph(cluster: any, reverse: boolean) {

      cluster = tree().nodeSize([20, 70])(cluster)

      const g = svg.append('g')
        .attr('id','lineage')
        .attr('font-family', 'sans-serif')
        .attr('font-size', 10)
        .attr('transform', `translate(${width/2}, ${height/2})`)

      g.append('g')
        .attr('fill', 'none')
        .attr('stroke-width', strokeWidth)
        .selectAll('path')
        .data(cluster.links())
        .join('path')
        .attr('d', linkHorizontal().x((d: any) => reverse ? -d.y : d.y).y((d: any) => d.x))
        .attr('stroke', (d: any) => d.target.data.matches && d.source.data.matches ? linkHighlight : defaultHighlight)

      const datasets = _filter(cluster.descendants(), d => isDataset(d))
      const jobs = _filter(cluster.descendants(), d => !isDataset(d))

      const datasetNode = g.append('g')
        .attr('stroke-linejoin', 'round')
        .selectAll('g')
        .data(datasets)
        .join('g')
        .attr('transform', d => `translate(${reverse ? -d.y : d.y},${d.x})`)

      const jobNode = g
        .append('g')
        .attr('stroke-linejoin', 'round')
        .selectAll('g')
        .data(jobs)
        .join('g')
        .attr('transform', d => `translate(${reverse ? -d.y : d.y},${d.x})`)

      datasetNode
        .append('a')
        .attr('href', (d: any) => ('/datasets/' + d.data.name))
        .append('rect')
        .attr('fill', d => d.data.matches ? circleHighlight : defaultHighlight)
        .attr('x', -square/2)
        .attr('y', -square/2)
        .attr('width', square)
        .attr('height', square)

      jobNode
        .append('a')
        .attr('href', (d: any) => ('/jobs/' + d.data.name))
        .append('circle')
        .attr('fill', d => d.data.matches ? findJobColor(d) : defaultHighlight)
        .attr('r', radius)

      // Add text to nodes
      datasetNode.append('text')
        .text(d => d.data.matches ? d.data.name : null)
        .attr('dy', 10)
        .attr('font-size', 8)
        .attr('font-family', 'sans-serif')
        .attr('transform', `rotate(45) translate(${-(radius + 4)}, ${-radius})`)
        .attr('text-anchor', 'end')
        .attr('fill', labelHighlight)

      // Add text to nodes
      jobNode.append('text')
        .text(d => d.data.matches ? d.data.name : null)
        .attr('dy', 10)
        .attr('font-size', 8)
        .attr('font-family', 'sans-serif')
        .attr('transform', `rotate(45) translate(${-(radius + 4)}, ${-radius})`)
        .attr('text-anchor', 'end')
        .attr('fill', labelHighlight)

      return svg.node()
    }

    svg.call(
      zoom()
        .scaleExtent([.8, 1.2])
        .on("zoom", () => { svg.selectAll("#lineage").attr("transform", event.transform); })
    )

    // run calculations for network graph
    let lineages = getLineages()
    lineages = nodeId ? [_find(lineages, lineage => lineage.name == nodeId)] : lineages
    let clusters = _map(lineages, lineage => hierarchy(lineage))
    clusters = _sortBy(clusters, l => l.descendants().length)
    const largestCluster = clusters[clusters.length - 1]
    if (largestCluster) {
      const rootNode = largestCluster.data
      const reverseLineage = getReverseLineage(rootNode)
      const reverseCluster = hierarchy(reverseLineage)

      // remove svg elements
      svg.selectAll('*').remove()

      graph(largestCluster, false)
      graph(reverseCluster, true)
    }

    if (this.props.isLoading !== newProps.isLoading) {
      return true
    } else {
      return false
    }
  }

  graph: SVGElement

  render(): React.ReactElement {
    const { classes, isLoading } = this.props

    return (
      <div id='network-graph-container' className={classes.networkBackground}>
        <div id='tooltip' className={classes.tooltip}></div>
        <Legend customClassName={classes.legend}></Legend>
        {isLoading ? (
          <Loader />
        ) : (
          <svg id='network-graph' className={classes.networkBackground}>
            <g
              ref={node => {
                this.graph = node as SVGElement
              }}
            >
            </g>
          </svg>
        )}
      </div>
    )
  }
}

export default (withStyles(styles)(NetworkGraph))
