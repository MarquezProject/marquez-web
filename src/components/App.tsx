import React, { ReactElement } from 'react'
import { Helmet } from 'react-helmet'
import { CssBaseline } from '@material-ui/core'
import AppBar from './AppBar'
const globalStyles = require('../global_styles.css')
const { neptune, telescopeBlack } = globalStyles
import { Route, Switch } from 'react-router-dom'
import { Grid } from '@material-ui/core'
import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles'
import { createStore, applyMiddleware } from 'redux'
import { Provider } from 'react-redux'
import logger from 'redux-logger'
import createSagaMiddleware from 'redux-saga'
import { composeWithDevTools } from 'redux-devtools-extension'
import { createBrowserHistory } from 'history'
import { routerMiddleware, ConnectedRouter } from 'connected-react-router'
import {
  withStyles,
  createStyles,
  WithStyles as IWithStyles,
  Theme as ITheme
} from '@material-ui/core/styles'

import createRootReducer from '../reducers'
import rootSaga from '../sagas'
import HomeContainer from '../containers/HomeContainer'
import Toast from '../containers/ToastContainer'
import NetworkGraphContainer from '../containers/NetworkGraphContainer'
import DatasetsContainer from '../containers/DatasetsContainer'

const sagaMiddleware = createSagaMiddleware({
  onError: (error, _sagaStackIgnored) => {
    console.log('There was an error in the saga', error)
  }
})
const history = createBrowserHistory()
const historyMiddleware = routerMiddleware(history)

const middleware = __DEVELOPMENT__
  ? [sagaMiddleware, historyMiddleware, logger]
  : [sagaMiddleware, historyMiddleware]

const store = createStore(
  createRootReducer(history),
  composeWithDevTools(applyMiddleware(...middleware))
)

sagaMiddleware.run(rootSaga)
// see the defaults here for MUI Theme -- https://material-ui.com/customization/default-theme/?expend-path=$.typography
const theme = createMuiTheme({
  typography: {
    h3: {
      fontSize: '1rem',
      fontWeight: 700,
      lineHeight: 2
    },
    fontSize: 14
  },
  palette: {
    primary: {
      main: telescopeBlack
    },
    secondary: {
      main: neptune
    }
  }
})

const styles = (_theme: ITheme) => {
  return createStyles({
    root: {
      height: '100vh',
      display: 'flex'
    }
  })
}

interface IProps extends IWithStyles<typeof styles> {}

const TITLE = 'Marquez | Data Kit'

const App = ({ classes }: IProps): ReactElement => {
  return (
    <Provider store={store}>
      <ConnectedRouter history={history}>
        <MuiThemeProvider theme={theme}>
          <Helmet>
            <title>{TITLE}</title>
          </Helmet>
          <CssBaseline />
          <Grid direction='column' alignItems='stretch' classes={classes} justify='space-between'>
            <AppBar />
            <NetworkGraphContainer />
            <Switch>
              <Route path='/' exact component={HomeContainer} />
              <Route path='/datasets/:name' exact component={DatasetsContainer} />
            </Switch>
            <Toast />
          </Grid>
        </MuiThemeProvider>
      </ConnectedRouter>
    </Provider>
  )
}

export default withStyles(styles)(App)
