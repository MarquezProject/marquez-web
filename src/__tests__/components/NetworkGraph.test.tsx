import { mount } from 'enzyme'
import * as React from 'react'
import { NetworkGraph } from '../../components/NetworkGraph'
const jobs = require('../../../docker/db/data/jobs.json')
const datasets = require('../../../docker/db/data/datasets.json')

describe('NetworkGraph Component', () => {
  // const wrapper = mount(<NetworkGraph jobs={jobs} datasets={datasets} />)
  test.skip('Should render', () => {
    expect(wrapper.exists()).toBe(true)
  })
  test.skip('renders a snapshot that matches previous', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
