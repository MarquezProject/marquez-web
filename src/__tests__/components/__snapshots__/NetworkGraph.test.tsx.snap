// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkGraph Component renders a snapshot that matches previous 1`] = `
<WithStyles(NetworkGraph)
  datasets={
    Array [
      Object {
        "description": "describes different geometries in the galaxy",
        "name": "galaxy.geometry",
        "namespace": "galactic-data",
        "physicalName": "galaxy.geometry",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all possible dimensions in the galaxy",
        "name": "galaxy.dimensions",
        "namespace": "galactic-data",
        "physicalName": "galaxy.dimensions",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "all trips taken through the galaxy",
        "name": "galaxy.trips",
        "namespace": "galactic-data",
        "physicalName": "galaxy.trips",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all possible dimensions on earth",
        "name": "earth.dimensions",
        "namespace": "galactic-data",
        "physicalName": "earth.dimensions",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all possible trips taken on earth",
        "name": "earth.trips",
        "namespace": "galactic-data",
        "physicalName": "earth.trips",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all stars in the galaxy",
        "name": "objects.stars",
        "namespace": "galactic-data",
        "physicalName": "objects.stars",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all planets in the galaxy",
        "name": "objects.planets",
        "namespace": "galactic-data",
        "physicalName": "objects.planets",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all life",
        "name": "objects.life",
        "namespace": "galactic-data",
        "physicalName": "objects.life",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes the entire universe",
        "name": "environment.universe",
        "namespace": "galactic-data",
        "physicalName": "environment.universe",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all radiowaves",
        "name": "galaxy.radiowaves",
        "namespace": "galactic-data",
        "physicalName": "galaxy.radiowaves",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all satellites",
        "name": "galaxy.satellites",
        "namespace": "galactic-data",
        "physicalName": "galaxy.satellites",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all conversations in the galaxy",
        "name": "environment.conversations",
        "namespace": "galactic-data",
        "physicalName": "environment.conversations",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "describes all transcripts in the galaxy",
        "name": "environment.transcripts",
        "namespace": "galactic-data",
        "physicalName": "environment.transcripts",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
      Object {
        "description": "this dataset is never used by a job!",
        "name": "environment.empty",
        "namespace": "galactic-data",
        "physicalName": "environment.empty",
        "sourceName": "redshift",
        "type": "DB_TABLE",
      },
    ]
  }
  jobs={
    Array [
      Object {
        "description": "calculates all possible dimensions from input geometry",
        "inputs": Array [
          "galaxy.geometry",
        ],
        "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_dimensions.sql",
        "name": "calculate.galaxy_dimensions",
        "namespace": "galactic-data",
        "outputs": Array [
          "galaxy.dimensions",
        ],
        "type": "BATCH",
      },
      Object {
        "description": "calculates trips taken in the galaxy",
        "inputs": Array [
          "galaxy.dimensions",
        ],
        "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
        "name": "calculate.galaxy_trips",
        "namespace": "galactic-data",
        "outputs": Array [
          "galaxy.trips",
        ],
        "type": "BATCH",
      },
      Object {
        "description": "calculates the dimensions of earth",
        "inputs": Array [
          "earth.dimensions",
        ],
        "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
        "name": "calculate.earth_dimensions",
        "namespace": "galactic-data",
        "outputs": Array [
          "earth.trips",
        ],
        "type": "BATCH",
      },
      Object {
        "description": "calculates the entire universe",
        "inputs": Array [
          "objects.stars",
          "objects.planets",
          "objects.life",
        ],
        "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
        "name": "calculate.universe",
        "namespace": "galactic-data",
        "outputs": Array [
          "environment.universe",
        ],
        "type": "BATCH",
      },
      Object {
        "description": "calculates all communications in the galaxy",
        "inputs": Array [
          "galaxy.radiowaves",
          "galaxy.satellites",
        ],
        "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
        "name": "calculate.communications",
        "namespace": "galactic-data",
        "outputs": Array [
          "environment.conversations",
          "environment.transcripts",
        ],
        "type": "BATCH",
      },
      Object {
        "description": "calculates literally nothing!",
        "inputs": Array [],
        "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
        "name": "calculate.nothing",
        "namespace": "galactic-data",
        "outputs": Array [],
        "type": "BATCH",
      },
    ]
  }
>
  <NetworkGraph
    classes={
      Object {
        "legend": "NetworkGraph-legend-3",
        "networkBackground": "NetworkGraph-networkBackground-1",
        "tooltip": "NetworkGraph-tooltip-2",
      }
    }
    datasets={
      Array [
        Object {
          "description": "describes different geometries in the galaxy",
          "name": "galaxy.geometry",
          "namespace": "galactic-data",
          "physicalName": "galaxy.geometry",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all possible dimensions in the galaxy",
          "name": "galaxy.dimensions",
          "namespace": "galactic-data",
          "physicalName": "galaxy.dimensions",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "all trips taken through the galaxy",
          "name": "galaxy.trips",
          "namespace": "galactic-data",
          "physicalName": "galaxy.trips",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all possible dimensions on earth",
          "name": "earth.dimensions",
          "namespace": "galactic-data",
          "physicalName": "earth.dimensions",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all possible trips taken on earth",
          "name": "earth.trips",
          "namespace": "galactic-data",
          "physicalName": "earth.trips",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all stars in the galaxy",
          "name": "objects.stars",
          "namespace": "galactic-data",
          "physicalName": "objects.stars",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all planets in the galaxy",
          "name": "objects.planets",
          "namespace": "galactic-data",
          "physicalName": "objects.planets",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all life",
          "name": "objects.life",
          "namespace": "galactic-data",
          "physicalName": "objects.life",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes the entire universe",
          "name": "environment.universe",
          "namespace": "galactic-data",
          "physicalName": "environment.universe",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all radiowaves",
          "name": "galaxy.radiowaves",
          "namespace": "galactic-data",
          "physicalName": "galaxy.radiowaves",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all satellites",
          "name": "galaxy.satellites",
          "namespace": "galactic-data",
          "physicalName": "galaxy.satellites",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all conversations in the galaxy",
          "name": "environment.conversations",
          "namespace": "galactic-data",
          "physicalName": "environment.conversations",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "describes all transcripts in the galaxy",
          "name": "environment.transcripts",
          "namespace": "galactic-data",
          "physicalName": "environment.transcripts",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
        Object {
          "description": "this dataset is never used by a job!",
          "name": "environment.empty",
          "namespace": "galactic-data",
          "physicalName": "environment.empty",
          "sourceName": "redshift",
          "type": "DB_TABLE",
        },
      ]
    }
    jobs={
      Array [
        Object {
          "description": "calculates all possible dimensions from input geometry",
          "inputs": Array [
            "galaxy.geometry",
          ],
          "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_dimensions.sql",
          "name": "calculate.galaxy_dimensions",
          "namespace": "galactic-data",
          "outputs": Array [
            "galaxy.dimensions",
          ],
          "type": "BATCH",
        },
        Object {
          "description": "calculates trips taken in the galaxy",
          "inputs": Array [
            "galaxy.dimensions",
          ],
          "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
          "name": "calculate.galaxy_trips",
          "namespace": "galactic-data",
          "outputs": Array [
            "galaxy.trips",
          ],
          "type": "BATCH",
        },
        Object {
          "description": "calculates the dimensions of earth",
          "inputs": Array [
            "earth.dimensions",
          ],
          "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
          "name": "calculate.earth_dimensions",
          "namespace": "galactic-data",
          "outputs": Array [
            "earth.trips",
          ],
          "type": "BATCH",
        },
        Object {
          "description": "calculates the entire universe",
          "inputs": Array [
            "objects.stars",
            "objects.planets",
            "objects.life",
          ],
          "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
          "name": "calculate.universe",
          "namespace": "galactic-data",
          "outputs": Array [
            "environment.universe",
          ],
          "type": "BATCH",
        },
        Object {
          "description": "calculates all communications in the galaxy",
          "inputs": Array [
            "galaxy.radiowaves",
            "galaxy.satellites",
          ],
          "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
          "name": "calculate.communications",
          "namespace": "galactic-data",
          "outputs": Array [
            "environment.conversations",
            "environment.transcripts",
          ],
          "type": "BATCH",
        },
        Object {
          "description": "calculates literally nothing!",
          "inputs": Array [],
          "location": "https://github.com/Galaxy/blob/ecfc04de65137bfd25a0416d227eaba5325c8c63/calculate.galaxy_trips.sql",
          "name": "calculate.nothing",
          "namespace": "galactic-data",
          "outputs": Array [],
          "type": "BATCH",
        },
      ]
    }
  >
    <div>
      <div
        className="NetworkGraph-tooltip-2"
        id="tooltip"
      />
      <svg
        className="NetworkGraph-networkBackground-1"
        id="network-graph"
      >
        <g>
          <g
            id="links"
          />
          <g
            id="jobNodes"
          />
          <g
            id="datasetNodes"
          />
        </g>
      </svg>
      <WithStyles(Legend)
        customClassName="NetworkGraph-legend-3"
      >
        <Legend
          classes={
            Object {
              "datasetShape": "Legend-datasetShape-4",
              "jobShape": "Legend-jobShape-5",
              "shape": "Legend-shape-6",
              "text": "Legend-text-7",
            }
          }
          customClassName="NetworkGraph-legend-3"
        >
          <Styled(MuiBox)
            className="NetworkGraph-legend-3"
            display="flex"
          >
            <div
              className="MuiBox-root MuiBox-root-9 NetworkGraph-legend-3"
            >
              <div
                className="Legend-datasetShape-4 Legend-shape-6"
              />
              <WithStyles(ForwardRef(Typography))
                className="Legend-text-7"
              >
                <ForwardRef(Typography)
                  className="Legend-text-7"
                  classes={
                    Object {
                      "alignCenter": "MuiTypography-alignCenter",
                      "alignJustify": "MuiTypography-alignJustify",
                      "alignLeft": "MuiTypography-alignLeft",
                      "alignRight": "MuiTypography-alignRight",
                      "body1": "MuiTypography-body1",
                      "body2": "MuiTypography-body2",
                      "button": "MuiTypography-button",
                      "caption": "MuiTypography-caption",
                      "colorError": "MuiTypography-colorError",
                      "colorInherit": "MuiTypography-colorInherit",
                      "colorPrimary": "MuiTypography-colorPrimary",
                      "colorSecondary": "MuiTypography-colorSecondary",
                      "colorTextPrimary": "MuiTypography-colorTextPrimary",
                      "colorTextSecondary": "MuiTypography-colorTextSecondary",
                      "displayBlock": "MuiTypography-displayBlock",
                      "displayInline": "MuiTypography-displayInline",
                      "gutterBottom": "MuiTypography-gutterBottom",
                      "h1": "MuiTypography-h1",
                      "h2": "MuiTypography-h2",
                      "h3": "MuiTypography-h3",
                      "h4": "MuiTypography-h4",
                      "h5": "MuiTypography-h5",
                      "h6": "MuiTypography-h6",
                      "noWrap": "MuiTypography-noWrap",
                      "overline": "MuiTypography-overline",
                      "paragraph": "MuiTypography-paragraph",
                      "root": "MuiTypography-root",
                      "srOnly": "MuiTypography-srOnly",
                      "subtitle1": "MuiTypography-subtitle1",
                      "subtitle2": "MuiTypography-subtitle2",
                    }
                  }
                >
                  <p
                    className="MuiTypography-root Legend-text-7 MuiTypography-body1"
                  >
                    datasets
                  </p>
                </ForwardRef(Typography)>
              </WithStyles(ForwardRef(Typography))>
              <div
                className="Legend-jobShape-5 Legend-shape-6"
              />
              <WithStyles(ForwardRef(Typography))
                className="Legend-text-7"
              >
                <ForwardRef(Typography)
                  className="Legend-text-7"
                  classes={
                    Object {
                      "alignCenter": "MuiTypography-alignCenter",
                      "alignJustify": "MuiTypography-alignJustify",
                      "alignLeft": "MuiTypography-alignLeft",
                      "alignRight": "MuiTypography-alignRight",
                      "body1": "MuiTypography-body1",
                      "body2": "MuiTypography-body2",
                      "button": "MuiTypography-button",
                      "caption": "MuiTypography-caption",
                      "colorError": "MuiTypography-colorError",
                      "colorInherit": "MuiTypography-colorInherit",
                      "colorPrimary": "MuiTypography-colorPrimary",
                      "colorSecondary": "MuiTypography-colorSecondary",
                      "colorTextPrimary": "MuiTypography-colorTextPrimary",
                      "colorTextSecondary": "MuiTypography-colorTextSecondary",
                      "displayBlock": "MuiTypography-displayBlock",
                      "displayInline": "MuiTypography-displayInline",
                      "gutterBottom": "MuiTypography-gutterBottom",
                      "h1": "MuiTypography-h1",
                      "h2": "MuiTypography-h2",
                      "h3": "MuiTypography-h3",
                      "h4": "MuiTypography-h4",
                      "h5": "MuiTypography-h5",
                      "h6": "MuiTypography-h6",
                      "noWrap": "MuiTypography-noWrap",
                      "overline": "MuiTypography-overline",
                      "paragraph": "MuiTypography-paragraph",
                      "root": "MuiTypography-root",
                      "srOnly": "MuiTypography-srOnly",
                      "subtitle1": "MuiTypography-subtitle1",
                      "subtitle2": "MuiTypography-subtitle2",
                    }
                  }
                >
                  <p
                    className="MuiTypography-root Legend-text-7 MuiTypography-body1"
                  >
                    jobs
                  </p>
                </ForwardRef(Typography)>
              </WithStyles(ForwardRef(Typography))>
            </div>
          </Styled(MuiBox)>
        </Legend>
      </WithStyles(Legend)>
    </div>
  </NetworkGraph>
</WithStyles(NetworkGraph)>
`;
